### PR TITLE
Top n push down

### PR DESCRIPTION
**Pull request by Novemser**
Support for `Limit` and `order by` clause.

### CODE REVIEW

**Comment by Novemser** [comment]
Refer to client PR: [#178](https://github.com/pingcap/tikv-client-lib-java/pull/178)

File: src/main/scala/org/apache/spark/sql/TiStrategy.scala [hunk 1]
```diff
@@ -146,6 +147,70 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
...
+    // If sortOrder is not null, limit must be greater than 0
+    if (limit < 0 || (sortOrder == null && limit == 0)) {
+      return Nil
+    }
+
+    child match {
+      case PhysicalOperation(projectList, filters, LogicalRelation(source: TiDBRelation, _, _))
+          if filters.forall(TiUtils.isSupportedFilter(_, source)) =>
+        pruneTopNFilterProject(limit, projectList, filters, source) :: Nil
+      case TiAggregation(
```
Comment (ilovesoup) [inline src/main/scala/org/apache/spark/sql/TiStrategy.scala]: Would you please confirm if topN and aggregates can be pushdown independently. 
For example, if you have group by name and count
you have 
region 1: A, 100, B, 1
region 2: A, 100, B 1000
and you take Top 1. In above case, B will be ignored in each region request but actually B is the group to take.
I wonder if above case works in your implementation.

File: src/main/scala/org/apache/spark/sql/TiStrategy.scala [hunk 2]
```diff
@@ -225,8 +266,126 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
...
+
+    aggregate.AggUtils.planAggregateWithoutDistinct(
+      groupingExpressions,
+      residualAggregateExpressions,
+      rewrittenResultExpression,
+      toCoprocessorRDD(source, output, dagReq)
+    )
+  }
+
+  def validGroupAggregate(groupingExpressions: Seq[NamedExpression],
```
Comment (ilovesoup) [inline src/main/scala/org/apache/spark/sql/TiStrategy.scala]: Change it to isValidAggregates. Aggregates include aggexpr and groupby exprs.

File: src/main/scala/org/apache/spark/sql/TiStrategy.scala [hunk 3]
```diff
@@ -248,89 +407,15 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
...
-        val output = (pushdownAggregates.map(x => toAlias(x)) ++ groupingExpressions)
-          .map(_.toAttribute)
-
-        val projectSeq: Seq[Attribute] = projects.asInstanceOf[Seq[Attribute]]
-        projectSeq.foreach(attr => dagReq.addRequiredColumn(TiColumnRef.create(attr.name)))
-
-        aggregate.AggUtils.planAggregateWithoutDistinct(
+          ) if validGroupAggregate(groupingExpressions, aggregateExpressions, source) =>
+        val dagReq: TiDAGRequest = filterToDAGRequest(filters, source)
+        groupAggregateProjection(
```
Comment (ilovesoup) [inline src/main/scala/org/apache/spark/sql/TiStrategy.scala]: same here

File: src/main/scala/org/apache/spark/sql/TiStrategy.scala (hunk 2 above)
Comment (Novemser) [reply src/main/scala/org/apache/spark/sql/TiStrategy.scala]: Changed.

File: src/main/scala/org/apache/spark/sql/TiStrategy.scala (hunk 3 above)
Comment (Novemser) [reply src/main/scala/org/apache/spark/sql/TiStrategy.scala]: Done.

**Comment by Novemser** [comment]
@ilovesoup Sorry for the inconvenience I bought, this PR only handles TopN logic but not aggregation with TopN logic.

**Comment by Novemser** [comment]
Integration test result:
Tests run: 7211  Tests succeeded: 3735  Tests failed: 21  Tests skipped: 3455
Result is as expected.

**Pull request merged**